openfasoc/generators/curref-gen/generators/custom_functions.py

```diff
@@ -0,0 +1,821 @@
+from decimal import Decimal
+from typing import Literal, Optional, Union
+
+from gdsfactory.component import Component
+from gdsfactory.components.rectangle import rectangle
+# Imports for tap ring
+from gdsfactory.components.rectangular_ring import rectangular_ring
+from gdsfactory.functions import transformed
+from glayout.flow.pdk.mappedpdk import MappedPDK
+from glayout.flow.pdk.sky130_mapped import sky130_mapped_pdk as sky130
+from glayout.flow.pdk.util.comp_utils import (
+    add_ports_perimeter,
+    align_comp_to_port,
+    evaluate_bbox,
+    prec_ref_center,
+    to_decimal,
+    to_float,
+)
+from glayout.flow.pdk.util.port_utils import (
+    rename_ports_by_list,
+    rename_ports_by_orientation,
+)
+from glayout.flow.pdk.util.snap_to_grid import component_snap_to_grid
+from glayout.flow.primitives.fet import multiplier
+from glayout.flow.primitives.via_gen import via_array, via_stack
+from glayout.flow.routing.c_route import c_route
+from glayout.flow.routing.L_route import L_route
+from glayout.flow.routing.straight_route import straight_route
+
+
+def macro_two_transistor_placement_Onchip(
+    pdk: MappedPDK,
+    deviceA_and_B: Literal["nfet", "pfet"],
+    with_substrate_tap: bool = False,
+    with_tie: bool = True,
+    width1: float = 1,
+    width2: float = 1,
+    length1: float = None,
+    length2: float = None,
+    fingers1: int = 3,
+    fingers2: int = 3,
+    matriz: list = [[0,2,1,1,2,0],[0,1,2,2,1,0],[0,2,1,1,2,0],[0,1,2,2,1,0]],
```

Comment:
These matrices seem to repeat a lot. Can they be converted into constants or a function with some documentation?